### PR TITLE
[CoSimulationApplication] 3D domain (volume) - 1D domain (line) data transfer operator

### PR DESCRIPTION
**Pull request by loumalouomega**
**📝 Description**

This is a very "not processional" PR, I don't like the quality of the result, and I was doing this in a private development, but due to certain requirements in @KratosMultiphysics/altair, I am uploading it to the main repository.

This adds  `DataTransfer3D1DProcess` and the corresponding *CoSimumlation data transfer operator*, in order to transfer from a 1D domain to a 3D domain, working similarly to a mapper. In the 3D-1D works the same as a mapper. In order to do so, it creates an auxiliary intersected modelpart between the 3D and 1D domain, then uses the mapper.

Originally was purely tetra-line, but I replaced the ` IntersectionUtilities` from #10463 with the `IsInside` from geometries. This is more generic and robust, but in some corner cases (line intersect only one corner) could give a bit worse results. In the future, I will create a meta-mapper like the 2D-3D, which will replace current implementation, as this requires extra work for MPI, and we need it now I do it simply as a data transfer operator.

Includes #10463 (despite is not needed any more). 

The tests included:

- **Block:**

image

- **Torus:**

image

**🆕 Changelog**

- Added #10463
- Added `DataTransfer3D1DProcess`
- Added tests

### CODE REVIEW

**Review by philbucher — CHANGES_REQUESTED** [review]
Before reviewing this in detail, please clean it up a bit

The size of the test needs to be reduced (best to make a new clean pr so that this does not end up in the enormous git-history)

Also please separate the core changes

**Comment by loumalouomega** [comment]
> Before reviewing this in detail, please clean it up a bit
> 
> The size of the test needs to be reduced (best to make a new clean pr so that this does not end up in the enormous git-history)
> 
> Also please separate the core changes

Okay

File: applications/CoSimulationApplication/tests/test_3d_1d_data_transfer_process.py [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+# Import the test utilities
+from KratosMultiphysics.from_json_check_result_process import FromJsonCheckResultProcess
+from KratosMultiphysics.json_output_process import JsonOutputProcess
+
+# Import basic dependencies
+import math
+import os
+
+def GetFilePath(fileName):
+    return os.path.join(os.path.dirname(os.path.realpath(__file__)), fileName)
```
Comment (philbucher) [inline applications/CoSimulationApplication/tests/test_3d_1d_data_transfer_process.py]: `pathlib.Path`

File: applications/CoSimulationApplication/tests/test_3d_1d_data_transfer_process.py [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+        result_file = GetFilePath("3d_1d_data_transfer/3d_1d_data_transfer_solutions/block_from_1d_to_3d")
+
+        # Check results
+        # generate_result(self.current_model, result_file, "Block")
+        check_results(self.current_model, result_file, "Block")
+
+        # # Debug
+        debug_vtk(self.current_model, ["Block", "Line"])
+
+    @KratosUnittest.skipIf(KM.IsDistributedRun(), "This test is designed for serial runs only.")
```
Comment (philbucher) [inline applications/CoSimulationApplication/tests/test_3d_1d_data_transfer_process.py]:
```suggestion
```

those are not needed, we never use this in other tests

File: applications/CoSimulationApplication/tests/test_3d_1d_data_transfer_process.py [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+        del node
+
+        for node in self.model_part_circle.Nodes:
+            node.SetSolutionStepValue(KM.TEMPERATURE,   node.X - node.Y)
+        del node
+
+    def tearDown(self):
+        pass
+
+    @KratosUnittest.skipIf(KM.IsDistributedRun(), "This test is designed for serial runs only.")
```
Comment (philbucher) [inline applications/CoSimulationApplication/tests/test_3d_1d_data_transfer_process.py]:
```suggestion
```

File: applications/CoSimulationApplication/tests/test_3d_1d_data_transfer_process.py [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+        result_file = GetFilePath("3d_1d_data_transfer/3d_1d_data_transfer_solutions/torus_from_1d_to_3d")
+
+        # Check results
+        # generate_result(self.current_model, result_file, "Torus")
+        check_results(self.current_model, result_file, "Torus")
+
+        # # Debug
+        debug_vtk(self.current_model, ["Torus", "Circle"])
+
+    @KratosUnittest.skipIf(KM.IsDistributedRun(), "This test is designed for serial runs only.")
```
Comment (philbucher) [inline applications/CoSimulationApplication/tests/test_3d_1d_data_transfer_process.py]:
```suggestion
```

File: applications/CoSimulationApplication/tests/test_3d_1d_data_transfer_process.py [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+        del node
+
+        for node in self.model_part_line.Nodes:
+            node.SetSolutionStepValue(KM.TEMPERATURE, - node.Z)
+        del node
+
+    def tearDown(self):
+        pass
+
+    @KratosUnittest.skipIf(KM.IsDistributedRun(), "This test is designed for serial runs only.")
```
Comment (philbucher) [inline applications/CoSimulationApplication/tests/test_3d_1d_data_transfer_process.py]:
```suggestion
```

File: applications/CoSimulationApplication/custom_processes/data_transfer_3D_1D_process.cpp [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+double DataTransfer3D1DProcess::GetMaxLength(ModelPart& rModelPart)
+{
+    auto& r_elements_array = rModelPart.Elements();
+    KRATOS_ERROR_IF(r_elements_array.size() == 0) << "Empty model part" << std::endl;
+    return block_for_each<MaxReduction<double>>(r_elements_array, [&](Element& rElement) {
```
Comment (philbucher) [inline applications/CoSimulationApplication/custom_processes/data_transfer_3D_1D_process.cpp]:
```suggestion
    return block_for_each<MaxReduction<double>>(r_elements_array, [&](const auto& rElement) {
```

File: applications/CoSimulationApplication/custom_processes/data_transfer_3D_1D_process.h [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+        std::vector<const Variable<double>*>& rOriginListVariables,
+        std::vector<const Variable<double>*>& rDestinationListVariables
+        );
+
+    /**
+     * @brief This method computes maximum length of the elements
+     * @param rModelPart The model part to compute
+     * @return The maximum length
+     */
+    static double GetMaxLength(ModelPart& rModelPart);
```
Comment (philbucher) [inline applications/CoSimulationApplication/custom_processes/data_transfer_3D_1D_process.h]:
```suggestion
    static double GetMaxLength(const ModelPart& rModelPart);
```

File: applications/CoSimulationApplication/custom_processes/data_transfer_3D_1D_process.cpp [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+        mpMapper->Map(*mOriginListVariables[i_var], *mDestinationListVariables[i_var], mapper_flags);
+    }
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+double DataTransfer3D1DProcess::GetMaxLength(ModelPart& rModelPart)
+{
+    auto& r_elements_array = rModelPart.Elements();
```
Comment (philbucher) [inline applications/CoSimulationApplication/custom_processes/data_transfer_3D_1D_process.cpp]: what about conditions?

Conditions have priority over elements in the mappingApp

File: applications/CoSimulationApplication/custom_processes/data_transfer_3D_1D_process.cpp [hunk 9]
```diff
@@ -0,0 +1,290 @@
...
+    const double max_length_1d = GetMaxLength(mr1DModelPart);
+    const double max_length_3d = GetMaxLength(mr3DModelPart);
+    const double max_length = std::max(max_length_1d, max_length_3d);
+
+    // The elements array
+    auto& r_elements_array = mr1DModelPart.Elements();
+    const auto it_elem_begin = r_elements_array.begin();
+
+    /// KDtree definitions
+    KRATOS_KDTREE_POINTELEMENT_DEFINITION
```
Comment (philbucher) [inline applications/CoSimulationApplication/custom_processes/data_transfer_3D_1D_process.cpp:141]: what does this do?

File: applications/CoSimulationApplication/tests/test_3d_1d_data_transfer_process.py [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+        # debug_vtk(self.current_model, ["Torus", "Circle"])
+
+def debug_vtk(model, list_names):
+    import KratosMultiphysics.vtk_output_process as vtk_output_process
+
+    for name in list_names:
+        vtk_output_parameters = KratosMultiphysics.Parameters("""{
+            "Parameters" : {
+                "model_part_name"                    : "",
+                "file_format"                        : "ascii",
```
Comment (philbucher) [inline applications/CoSimulationApplication/tests/test_3d_1d_data_transfer_process.py]:
```suggestion
```

File: applications/CoSimulationApplication/python_scripts/data_transfer_operators/3d_1d_data_transfer.py [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+            if self.origin_is_3d:
+                self.__data_transfer_process[identifier_tuple].Set(KM.Mapper.USE_TRANSPOSE)
+            self.__data_transfer_process[identifier_tuple].Execute()
+            if self.origin_is_3d:
+                self.__data_transfer_process[identifier_tuple].Reset(KM.Mapper.USE_TRANSPOSE)
+
+    def _Check(self, from_solver_data, to_solver_data):
+        def CheckData(data_to_check):
+            if "node" not in data_to_check.location:
+                raise Exception('Transfer only supports nodal values!"{}"\nChecking ModelPart "{}" of solver "{}"'.format(self._ClassName(), data_to_check.model_part_name, data_to_check.solver_name))
```
Comment (philbucher) [inline applications/CoSimulationApplication/python_scripts/data_transfer_operators/3d_1d_data_transfer.py]: please use f-strings in new code

File: applications/CoSimulationApplication/python_scripts/data_transfer_operators/3d_1d_data_transfer.py [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+
+        CheckData(from_solver_data)
+        CheckData(to_solver_data)
+
+    def __check_model_part_3D(self, model_part):
```
Comment (philbucher) [inline applications/CoSimulationApplication/python_scripts/data_transfer_operators/3d_1d_data_transfer.py:213]: what if the ModelPart has 1D and 3D elements?

File: applications/CoSimulationApplication/python_scripts/data_transfer_operators/3d_1d_data_transfer.py [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+            parameters["destination_variables"].Append(variable_destination.Name())
+            for transfer_option in transfer_options.GetStringArray():
+                if transfer_option == "swap_sign":
+                    parameters["swap_sign"].SetBool(True)
+
+            # Check if the origin model part is 3D or 1D
+            if self.__check_model_part_3D(model_part_origin):
+                self.origin_is_3d = True
+            else:
+                self.origin_is_3d = False
```
Comment (philbucher) [inline applications/CoSimulationApplication/python_scripts/data_transfer_operators/3d_1d_data_transfer.py]:
```suggestion
            self.origin_is_3d = self.__check_model_part_3D(model_part_origin)
```

File: applications/CoSimulationApplication/custom_processes/data_transfer_3D_1D_process.cpp [hunk 14]
```diff
@@ -0,0 +1,290 @@
...
+    Parameters mapping_variables = mThisParameters["mapping_variables"];
+    mpMapper = MapperFactoryType::CreateMapper(r_aux_model_part_3D, mr1DModelPart, mapping_variables);
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+DataTransfer3D1DProcess::~DataTransfer3D1DProcess()
+{
+}
```
Comment (philbucher) [inline applications/CoSimulationApplication/custom_processes/data_transfer_3D_1D_process.cpp]:
```suggestion
```

File: applications/CoSimulationApplication/custom_processes/data_transfer_3D_1D_process.h [hunk 15]
```diff
@@ -0,0 +1,306 @@
...
+
+    ///Copy constructor  (not really required)
+    PointElement(const PointElement& rRHS):
+        BaseType(rRHS),
+        mpElement(rRHS.mpElement)
+    {
+    }
+
+    /// Destructor.
+    ~PointElement() override= default;
```
Comment (philbucher) [inline applications/CoSimulationApplication/custom_processes/data_transfer_3D_1D_process.h]:
```suggestion
```

unused code => remove

File: applications/CoSimulationApplication/custom_processes/data_transfer_3D_1D_process.h [hunk 16]
```diff
@@ -0,0 +1,306 @@
...
+                }
+            } else {
+                if (rFirstModelPart.ConditionsBegin()->GetGeometry().LocalSpaceDimension() == 1) {
+                    return rFirstModelPart;
+                }
+            }
+        }
+
+        // Second model part
+        if (rSecondModelPart.NumberOfElements() > 0 || rSecondModelPart.NumberOfConditions() > 0 ) {
```
Comment (philbucher) [inline applications/CoSimulationApplication/custom_processes/data_transfer_3D_1D_process.h]: conditions take precedence over elements in the mappingApp FYI

**Review by philbucher — COMMENTED** [review]
can you please reduce the size of the mdpa files?

File: applications/CoSimulationApplication/custom_processes/data_transfer_3D_1D_process.cpp (hunk 9 above)
Comment (loumalouomega) [reply applications/CoSimulationApplication/custom_processes/data_transfer_3D_1D_process.cpp:141]: It is a macro defined in the header

File: applications/CoSimulationApplication/python_scripts/data_transfer_operators/3d_1d_data_transfer.py (hunk 12 above)
Comment (loumalouomega) [reply applications/CoSimulationApplication/python_scripts/data_transfer_operators/3d_1d_data_transfer.py:213]: I will raise an exception

File: applications/CoSimulationApplication/tests/test_3d_1d_data_transfer_process.py [hunk 17]
```diff
@@ -0,0 +1,279 @@
...
+    out_parameters = KM.Parameters("""{
+        "output_variables"     : ["TEMPERATURE"],
+        "output_file_name"     : "",
+        "model_part_name"      : "",
+        "historical_value"     : true,
+        "time_frequency"       : 0.0
+    }""")
+
+    out_parameters["model_part_name"].SetString(domain)
+    out_parameters["output_file_name"].SetString(str(output_filename) + "_data_transfer.json")
```
Comment (philbucher) [inline applications/CoSimulationApplication/tests/test_3d_1d_data_transfer_process.py]:
```suggestion
    out_parameters["output_file_name"].SetString(f"{output_filename}_data_transfer.json")
```

File: applications/CoSimulationApplication/python_scripts/data_transfer_operators/3d_1d_data_transfer.py [hunk 18]
```diff
@@ -0,0 +1,276 @@
...
+            if first_geometry_type is None:
+                first_geometry_type = geom.LocalSpaceDimension()
+            elif geom.LocalSpaceDimension() != first_geometry_type:
+                raise ValueError("Inconsistent geometry types found in model_part.")
+
+        # Check if the element is 1D
+        if first_geometry_type == 1:
+            return False
+        else:
+            return True
```
Comment (philbucher) [inline applications/CoSimulationApplication/python_scripts/data_transfer_operators/3d_1d_data_transfer.py]:
```suggestion
        return first_geometry_type != 1
```

File: applications/CoSimulationApplication/python_scripts/data_transfer_operators/3d_1d_data_transfer.py [hunk 19]
```diff
@@ -0,0 +1,276 @@
...
+            True if all elements are 3D and have the same geometry type, False if any 
+            element is 1D.
+        Raises:
+        ------
+        ValueError
+            If elements in `model_part` have different geometry types, an exception is raised.
+        """
+        first_geometry_type = None
+
+        for elem in model_part.Elements:
```
Comment (philbucher) [inline applications/CoSimulationApplication/python_scripts/data_transfer_operators/3d_1d_data_transfer.py]: again, what about conditions?

File: applications/CoSimulationApplication/custom_processes/data_transfer_3D_1D_process.h [hunk 20]
```diff
@@ -0,0 +1,297 @@
...
+
+    /// Default constructor.
+    DataTransfer3D1DProcess(
+        ModelPart& rFirstModelPart,
+        ModelPart& rSecondModelPart,
+        Parameters ThisParameters = Parameters(R"({})")
+        );
+
+    /// Destructor.
+    ~DataTransfer3D1DProcess() override = default;
```
Comment (philbucher) [inline applications/CoSimulationApplication/custom_processes/data_transfer_3D_1D_process.h]:
```suggestion
```

File: applications/CoSimulationApplication/custom_processes/data_transfer_3D_1D_process.h [hunk 21]
```diff
@@ -0,0 +1,297 @@
...
+///@name Functions
+///@{
+
+/**
+ * @brief This method determines the 1D model part
+ * @param rFirstModelPart The first ModelPart
+ * @param rSecondModelPart The second ModelPart
+ * @return The 1D model part 
+ */
+ModelPart& Determine1DModelPart(ModelPart& rFirstModelPart, ModelPart& rSecondModelPart)
```
Comment (philbucher) [inline applications/CoSimulationApplication/custom_processes/data_transfer_3D_1D_process.h]: can you move the logic to the source file?

**Review by philbucher — DISMISSED** [review]
LGTM, left minor comments

File: applications/CoSimulationApplication/custom_processes/data_transfer_3D_1D_process.h (hunk 20 above)
Comment (loumalouomega) [reply applications/CoSimulationApplication/custom_processes/data_transfer_3D_1D_process.h]: Without this does not compile

File: applications/CoSimulationApplication/python_scripts/data_transfer_operators/3d_1d_data_transfer.py (hunk 19 above)
Comment (loumalouomega) [reply applications/CoSimulationApplication/python_scripts/data_transfer_operators/3d_1d_data_transfer.py]: True

File: applications/CoSimulationApplication/custom_processes/data_transfer_3D_1D_process.h (hunk 21 above)
Comment (loumalouomega) [reply applications/CoSimulationApplication/custom_processes/data_transfer_3D_1D_process.h]: Okay

**Comment by philbucher** [comment]
Seems that you messed up a merge?

**Comment by loumalouomega** [comment]
> Seems that you messed up a merge?

Fixed, my WSL does bizarre things with git

**Review by philbucher — APPROVED** [review]

**Pull request merged**